aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+      // NB: dropout_seed passed dummy 0, because it isn't actually used
+      // when dropout_state is defined.
+      auto dropout_p = train ? dropout : 0;
+      DropoutDescriptor dropout_desc;
+      dropout_desc.set(handle, dropout_p, dropout_state, 0);
+      return dropout_desc;
+    }
+  };
+
+  // RNNDescriptor
+
+  struct RNNDescriptorParams {
+    int64_t hidden_size;
+    int64_t num_layers;
+    cudnnDirectionMode_t bidirectional;
+    cudnnRNNMode_t mode;
+    cudnnDataType_t datatype;
+
+    cudnnRNNInputMode_t input_mode = CUDNN_LINEAR_INPUT;
+
+    int64_t num_directions() const {
+      return bidirectional ? 2 : 1;
+    }
+
+    void set_mode(int64_t fn_mode) {
+      switch (fn_mode) {
+        case CUDNN_RNN_RELU:
+          mode = CUDNN_RNN_RELU;
+          break;
+        case CUDNN_RNN_TANH:
+          mode = CUDNN_RNN_TANH;
+          break;
+        case CUDNN_LSTM:
+          mode = CUDNN_LSTM;
+          break;
+        case CUDNN_GRU:
+          mode = CUDNN_GRU;
+          break;
+        default:
+          throw std::runtime_error("unrecognized mode"); // TODO
+      }
+    }
+
+    void set_bidirectional(bool fn_bidirectional) {
+      bidirectional = fn_bidirectional ? CUDNN_BIDIRECTIONAL : CUDNN_UNIDIRECTIONAL;
+    }
+
+    RNNDescriptor descriptor(cudnnHandle_t handle, DropoutDescriptor&& dropout_desc) const {
+      RNNDescriptor rnn_desc;
+      rnn_desc.set(handle, hidden_size, num_layers, std::move(dropout_desc), input_mode, bidirectional, mode, datatype);
+      return rnn_desc;
+    }
+
+    RNNDescriptor descriptor(cudnnHandle_t handle) const {
+      DropoutDescriptor dropout_desc;
+      dropout_desc.set(handle, 0, {}, 0);
+      return descriptor(handle, std::move(dropout_desc));
+    }
+  };
+
+  // TensorDescriptor list
+
+  std::vector<TensorDescriptor> rnn_descriptor_sequence(const Tensor& tensor, IntList batch_sizes) {
+    std::vector<TensorDescriptor> descriptors(batch_sizes.size());
+    size_t i = 0;
+    for (auto batch_size : batch_sizes) {
+      // NB: The narrow is solely to adjust the batch size; to do it
+      // accurately we would have to adjust the start index as well,
+      // but the pointer location isn't actually used so we can skip it.
+      // NB: cuDNN RNN API has an undocumented requirement that all
+      // tensors have dimension 5.
+      descriptors[i].set(tensor.narrow(0, 0, batch_size), 5);
+      i++;
+    }
+    return descriptors;
+  }
+
+  std::vector<TensorDescriptor> rnn_descriptor(const Tensor& tensor, int64_t N) {
+    std::vector<TensorDescriptor> descriptors(N);
+    for (int64_t i = 0; i < N; i++) {
+      descriptors[i].set(tensor, 5);
+    }
+    return descriptors;
+  }
+
+  struct TensorDescriptorListParams {
+    IntList batch_sizes;
+    int64_t seq_length;
+    int64_t mini_batch;
+    int64_t inner_size;  // previously known as "input_size"
+    int64_t outer_size;  // only valid when !is_input_packed
```

Comment:
`inner_size` is simply whatever you want to call the dimension in the input size `seq_length x batch_size x inner_size` (which is why it was called "input size"; it's the input size minus the sequence length / batch size.) After seeing https://stackoverflow.com/questions/43624590/use-of-cudnn-rnn/46783842#46783842 it seems like it might not be a bad idea to just call this `vocab_size`; it's very WLM specific but it should give code readers the right idea.

Reply:
```
  // The best way to understand the meaning of the values stored in
  // this struct is to consider each of the possible ways our
  // input can be structured.
  //
  // Suppose you want to run RNN on the following variable
  // length inputs:
  //
  //    ABCD
  //    EF
  //    G
  //
  // (Let _ be padding when we have non-packed representations.)
  //
  // # Packed input (batch_sizes is non-empty)
  //
  //  vocab_size
  // +------+                    \
  // | A    |                    |
  // | E    |                    |
  // | G    | batch_sizes[0] = 3 |
  // +------+                    |
  // | B    |                    | batch_sizes_sum = 7
  // | F    | batch_sizes[1] = 2 |
  // +------+                    |
  // | C    | batch_sizes[2] = 1 |
  // +------+                    |
  // | D    | batch_sizes[3] = 1 |
  // +------+                    /
  //
  //              (seq_length = 4)
  //
  //    input.size() = batch_sizes_sum x vocab_size
  //
  // # Unpacked input (batch_first = false)
  //
  //  batch_size = 3
  // +-------+
  // | A E G |
  // | B F _ | seq_length = 4
  // | C _ _ |
  // | D _ _ |
  // +-------+
  //    ...    vocab_size
  // +-------+
  //
  //    input.size() = seq_length x batch_size x vocab_size
  //
  // # Unpacked input (batch_first = true)
  //
  //  seq_length = 4
  // +---------+
  // | A B C D |
  // | E F _ _ | batch_size = 3
  // | G _ _ _ |
  // +---------+
  //     ...     vocab_size
  // +---------+
  //
  //    input.size() = batch_size x seq_length x vocab_size
  //
```

Reply:
Renamed inner_size => vocab_size, outer_size => batch_sizes_sum

Reply:
`vocab_size` refers to the (unrelated) size of the vocabulary; `embedding_size` or `input_size` would be a clearer way to refer to the size of the channel dimension of the tensor being sent to an RNN.

Reply:
or `n_input_channels`

Reply:
OK, I guess we'll rename it back to `input_size` :)